### DISCUSSION TITLE
Overlay editor issues : losing overlays and with placing objects

### DISCUSSION THREAD

**Discussion by ssagerian** [AllskyTeam/allsky-modules#75] [General]
Some how last night, my overlays disappeared, nothing in the live view, and nothing show in the overlay editor. Ok, so I used the overlay editor to add back my text fields and my image. For most of the day, everything seemed fine, I was able to see my site remotely  and things were fine. Then, the original lost text objects reappeared and the new text objects had been displaced and re-scaled. With the overlay editor, I edited the text fields and my image objects. at one point, I moved things a bit too close to the edge of the image window and everything was gone again. The log file has the following..
an 15 19:14:55 homesky allsky[4340]: ERROR: Field 'Fan status ON' is outside of the image
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Field 'Temperature: 32.6 °F' is outside of the image
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Field 'Home Sky Image ' is outside of the image
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Field 'Image period NIGHT' is outside of the image
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Field 'Camera Gain: 1' is outside of the image
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Field 'Camera RPi HQ' is outside of the image
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Image 'compass-red.png' is outside the bounds of the main image.
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Cannot locate image random_temperature_humidity_plot.png.
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Cannot locate image random_temperature_humidity_plot.png.
Jan 15 19:14:55 homesky allsky[4340]: ERROR: Image 'temp_humidity_plot.png' is outside the bounds of the ma
Ok, so when I go back to the overlay editor, there are no text fields, the editor is empty..but yet I still get these errors..

pi@homesky:~/allsky/scripts $ tail -n 20 /var/log/allsky.log
Jan 15 19:24:11 homesky allsky[7054]: ERROR: Cannot locate image random_temperature_humidity_plot.png.
Jan 15 19:24:11 homesky allsky[7054]: ERROR: Cannot locate image random_temperature_humidity_plot.png.
Jan 15 19:24:11 homesky allsky[7054]: ERROR: Image 'temp_humidity_plot.png' is outside the bounds of the main image.
Jan 15 19:25:31 homesky allsky[6914]: STARTING EXPOSURE at: 2024-01-15 19:25:31   @ 5.8 sec
Jan 15 19:25:40 homesky allsky[6914]:   > Saving NIGHT image 'image-20240115192531.jpg'
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Image 15-01-2024 19:25:31' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Allsky Version v2023.05.01_03' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Exposure Time 5,766,827' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Humidity 59.4' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Dew heater status ON' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Fan status ON' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Temperature: 32.8 °F' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Home Sky Image ' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Image period NIGHT' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Camera Gain: 1' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Field 'Camera RPi HQ' is outside of the image
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Image 'compass-red.png' is outside the bounds of the main image.
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Cannot locate image random_temperature_humidity_plot.png.
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Cannot locate image random_temperature_humidity_plot.png.
Jan 15 19:25:44 homesky allsky[7513]: ERROR: Image 'temp_humidity_plot.png' is outside the bounds of the main image

opening the editor..

![image](https://github.com/AllskyTeam/allsky-modules/assets/3128943/00cb6364-d6c6-4409-8e24-704bcdbf441a)

**Comment by ssagerian** [comment]
So this behavior still continues. I see in the log the same errors (every minute). I go to the overlay editor, there are no objects in the grid field.  Nothing to move, nothing to edit.

**Comment by ssagerian** [comment]
Ok, discovered that the overlay.json file contained x,y values for locations of objects outside the bounds of the image size.

**Comment by ssagerian** [comment]
One last comment. Im theorizing that while in the overlay editor, I adjusted the scale of my browser window (or some portion of that) and the the overlay editor gets it's xy coordinates in an absolute method, rather than a relative to the actual image size. It assigned the xy coordinates of the object using the coordinates using the browser window xy values(?) not scaled to the editor window image. (?)
Later when  combining the image file and the overlay objects the combining method choked on adding those objects.

As it was, the live image display show no overlays and the allsky log had a series of error reports about objects outside the image (as reported earlier) and lastly the overlay editor showed nothing. 

I solved the problem by editing the overlay.json file, tlx, tly, x, y values to fit with my 1014 x 760 image size, and resaving the file. I then had to go back to the overlay editor, the objects now reappeared and i was able to relocate them. The live view now has overlays.

If anyone has a explanation as to how the overlay.json file got those wildly larger coordinate values, that would be useful.

**Comment by ssagerian** [comment]
ok, this is making me sad..
no sooner than I thought this was a minor issue than I when to view the live view this morning and the live view had stopped. The allsky log has the following ..
 Processing end of night data
allsky[10778]: ==========
allsky[10778]: === Starting daytime capture ===
allsky[10778]: ==========
allsky[10778]: -----
allsky[10778]: STARTING EXPOSURE at: 2024-01-17 06:46:44   @ 793.70 ms (0.79 sec)
allsky[10778]:  > Running: libcamera-still --output '/home/pi/allsky/tmp/image-20240117064644.jpg' --timeout 1 --nopreview --width 4056 --height 3040 --shutter 793700 --saturation 0.4 --sharpness 1.5 --quality 100
allsky[26903]: endOfNight.sh: ===== Posting twilight data
allsky[26903]: endOfNight.sh: ===== Generating Startrails
allsky[10778]: terminate called after throwing an instance of 'cv::Exception'
allsky[10778]:   what():  OpenCV(4.5.1) ../modules/core/src/matrix_iterator.cpp:72: error: (-215:Assertion failed) A.size == arrays[i0]->size in function 'init'
allsky[10663]: /home/pi/allsky/allsky.sh: line 274: 10778 Aborted                 "${ALLSKY_BIN}/${CAPTURE}" -cmd "${RPi_COMMAND_TO_USE}" -debuglevel "${ALLSKY_DEBUG_LEVEL}" -config "${ARGS_FILE}"
allsky[10663]: ***
allsky[10663]: *** After fixing, restart the allsky service. ***
allsky[10663]: ***
allsky[27023]: Upload output from '/home/pi/allsky/tmp/image.jpg:
allsky[27023]:    cp: '/home/pi/allsky/tmp/image.jpg' and '/home/pi/allsky/tmp//image.jpg' are the same file
allsky[10663]:      ***** AllSky Stopped *****
allsky[11415]:      ***** Starting AllSky *****
allsky[11543]: Upload output from '/home/pi/allsky/tmp/image.jpg:
allsky[11543]:    cp: '/home/pi/allsky/tmp/image.jpg' and '/home/pi/allsky/tmp//image.jpg' are the same file


After restarting the live view, now my live view has a scaled version of the overlays...
![Screenshot 2024-01-17 at 08-14-52 Liveview - WebUI](https://github.com/AllskyTeam/allsky-modules/assets/3128943/b3a1d61b-a728-4775-9018-20dd8689ad6c)

**Comment by Alex-developer** [comment]
Could you please send me the overlay.json file from ~/allsky/config/overlay/

I will investigate using your overlay file and see if I can work out whats going on with the overlays

Thanks

Alex

**Comment by ssagerian** [comment]
sorry to be such a whiner..
thanks for looking into this..

**Reply by Alex-developer** [reply]
No problem at all :-)

If you upload the file in Github I will take a look for you

Alex

**Comment by ssagerian** [comment]
[overlay.json](https://github.com/AllskyTeam/allsky-modules/files/13969468/overlay.json)

**Comment by ssagerian** [comment]
odd that some x,y values are floats?

**Comment by Alex-developer** [comment]
> [overlay.json](https://github.com/AllskyTeam/allsky-modules/files/13969468/overlay.json)

Thanks for That. I am away until tomorrow night so will have a look at it over the next few days

The floats values are calculated. tlx and tly should always be ints but x and y are derived so can end up as floats

Cheers

Alex

**Comment by ssagerian** [comment]
ok, so I must be the poster child for not reading the documentation. Most of my issues appear to be related to the way I setup the local web server. Im now down to 2 errors each day, on transition from night to day and day to night. I get the following error on night to day transitions..

`#00FF00`
**Jan 19 17:20:16 : ==========
Jan 19 17:20:16 : === Starting nighttime capture ===
Jan 19 17:20:16 : ==========
Jan 19 17:20:16 : -----
Jan 19 17:20:16 : STARTING EXPOSURE at: 2024-01-19 17:20:16   @ 1.2 sec
Jan 19 17:20:16 :  > Running: libcamera-still --output '/home/pi/allsky/tmp/image-20240119172016.jpg' --timeout 1 --nopreview --width 1014 --height 760 --shutter 1189207 --analoggain 1 --awb auto --saturation 0.4 --sharpness 1.5 --quality 100
Jan 19 17:20:20 : terminate called after throwing an instance of 'cv::Exception'
Jan 19 17:20:20 :   what():  OpenCV(4.5.1) ../modules/core/src/matrix_iterator.cpp:72: error: (-215:Assertion failed) A.size == arrays[i0]->size in function 'init'
Jan 19 17:20:20 : /home/pi/allsky/allsky.sh: line 274:  6615 Aborted                 "${ALLSKY_BIN}/${CAPTURE}" -cmd "${RPi_COMMAND_TO_USE}" -debuglevel "${ALLSKY_DEBUG_LEVEL}" -config "${ARGS_FILE}"
Jan 19 17:20:20 : ***
Jan 19 17:20:20 : *** After fixing, restart the allsky service. ***
Jan 19 17:20:20 : ***

Jan 20 06:46:29 : endOfNight.sh: ===== Posting twilight data
Jan 20 06:46:31 : endOfNight.sh: ===== Generating Startrails
Jan 20 06:46:35 :+1:  terminate called after throwing an instance of 'cv::Exception'
Jan 20 06:46:39 :   what():  OpenCV(4.5.1) ../modules/core/src/matrix_iterator.cpp:72: error: (-215:Assertion failed) A.size == arrays[i0]->size in function 'init'
Jan 20 06:46:39 : /home/pi/allsky/allsky.sh: line 274: 18693 Aborted                 "${ALLSKY_BIN}/${CAPTURE}" -cmd "${RPi_COMMAND_TO_USE}" -debuglevel "${ALLSKY_DEBUG_LEVEL}" -config "${ARGS_FILE}"
Jan 20 06:46:39 : ***
Jan 20 06:46:39 : *** After fixing, restart the allsky service. ***
Jan 20 06:46:39 : *****
```

`#FFFFFF`


The issue with the size of the overlay objects still exists. Here is the odd part..before transitioning from night to day, the image location, the text fields are all good, correct size and correct locations. After the error (mentioned above) the overlays are very small and clustered into the left upper corner . After the day to night transition (and error causing allsky to stop) the overlays are ok again..

You can see the image via the local website just prior to 17:20 PM CST and just before 6:30 AM CST via the allsky map.Plainfield, Illinois

**Comment by Alex-developer** [comment]
I have started work on some changes to the overlays so will look at this as part of it, I am starting with the fields outside of the image

**Comment by Alex-developer** [comment]
Fixes for the overlay added to the next release - See https://github.com/AllskyTeam/allsky/issues/3317
